File: classified/packages/plugin-autocoder/src/actions/create-project.ts
import {
  Action,
  ActionResult,
  HandlerCallback,
  type IAgentRuntime,
  Memory,
  State,
  type UUID,
  elizaLogger,
} from '@elizaos/core';
import { v4 as uuidv4 } from 'uuid';
import type { ProjectMetadata, ProjectType } from '../types/index';
import { FormsService } from '@elizaos/plugin-forms';

/**
 * Extract project name from user message
 */
function extractProjectName(text: string): string | null {
  // Look for quoted names
  const quotedMatch = text.match(/["']([^"']+)["']/);
  if (quotedMatch) {
    return quotedMatch[1];
  }

  // Look for patterns like "called X", "named X"
  const namedMatch = text.match(/(?:called|named)\s+(\S+)/i);
  if (namedMatch) {
    return namedMatch[1];
  }

  // Extract from specific patterns
  const patterns = [/create\s+(?:a\s+)?(?:new\s+)?(\S+)/i, /(?:build|make)\s+(?:a\s+)?(\S+)/i];

  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) {
      return match[1];
    }
  }

  return null;
}

/**
 * Extract project type from user message
 */
function extractProjectType(text: string): ProjectType {
  const lower = text.toLowerCase();

  if (lower.includes('plugin')) {
    return 'plugin';
  }
  if (lower.includes('agent') || lower.includes('bot')) {
    return 'agent';
  }
  if (lower.includes('workflow') || lower.includes('flow')) {
    return 'workflow';
  }
  if (lower.includes('mcp') || lower.includes('model context')) {
    return 'mcp';
  }
  if (lower.includes('app') || lower.includes('application')) {
    return 'full-stack';
  }

  // Default to plugin for ElizaOS
  return 'plugin';
}

function getFormTemplate(projectType: string) {
  switch (projectType) {
    case 'plugin':
      return {
        title: 'ElizaOS Plugin Project',
        description: "Let's gather information to create your plugin",
        fields: [
          {
            id: 'project_name',
            label: 'Project Name',
            type: 'text',
            placeholder: 'my-awesome-plugin',
            required: true,
            helpText: 'What would you like to name your plugin?',
          },
          {
            id: 'description',
            label: 'Description',
            type: 'textarea',
            placeholder: 'A plugin that does amazing things...',
            required: true,
            helpText: 'Describe what your plugin will do',
          },
          {
            id: 'requirements',
            label: 'Requirements',
            type: 'textarea',
            placeholder: 'List the key features and functionality...',
            required: true,
            helpText: 'What specific features or capabilities should this plugin have?',
          },
          {
            id: 'apis',
            label: 'External APIs (optional)',
            type: 'textarea',
            placeholder: 'OpenAI API, Discord API, etc.',
            required: false,
            helpText: 'List any external APIs or services this plugin will use',
          },
          {
            id: 'test_scenarios',
            label: 'Test Scenarios (optional)',
            type: 'textarea',
            placeholder: 'User asks for weather, plugin responds with...',
            required: false,
            helpText: 'Describe some test scenarios to validate the plugin works correctly',
          },
        ],
      };

    case 'agent':
      return {
        title: 'ElizaOS Agent Project',
        description: "Let's design your agent's personality and capabilities",
        fields: [
          {
            id: 'agent_name',
            label: 'Agent Name',
            type: 'text',
            placeholder: 'My Assistant',
            required: true,
            helpText: "What's your agent's name?",
          },
          {
            id: 'personality',
            label: 'Personality & Bio',
            type: 'textarea',
            placeholder: 'A helpful, friendly assistant who...',
            required: true,
            helpText: "Describe your agent's personality and background",
          },
          {
            id: 'capabilities',
            label: 'Capabilities',
            type: 'textarea',
            placeholder: 'Can help with coding, answer questions...',
            required: true,
            helpText: 'What should your agent be able to do?',
          },
          {
            id: 'plugins',
            label: 'Plugins to Include',
            type: 'textarea',
            placeholder: 'twitter, discord, knowledge...',
            required: false,
            helpText: 'Which ElizaOS plugins should this agent use?',
          },
        ],
      };

    case 'workflow':
      return {
        title: 'Workflow Project',
        description: 'Define your workflow automation',
        fields: [
          {
            id: 'workflow_name',
            label: 'Workflow Name',
            type: 'text',
            placeholder: 'data-processing-workflow',
            required: true,
          },
          {
            id: 'description',
            label: 'Description',
            type: 'textarea',
            placeholder: 'This workflow processes data and...',
            required: true,
          },
          {
            id: 'steps',
            label: 'Workflow Steps',
            type: 'textarea',
            placeholder: '1. Fetch data\n2. Process\n3. Store results',
            required: true,
            helpText: 'List the steps in your workflow',
          },
          {
            id: 'triggers',
            label: 'Triggers',
            type: 'textarea',
            placeholder: 'On schedule, webhook, manual...',
            required: false,
          },
        ],
      };

    default:
      return {
        title: 'New Project',
        description: 'Tell me about your project',
        fields: [
          {
            id: 'project_name',
            label: 'Project Name',
            type: 'text',
            required: true,
          },
          {
            id: 'description',
            label: 'Description',
            type: 'textarea',
            required: true,
          },
          {
            id: 'requirements',
            label: 'Requirements',
            type: 'textarea',
            required: true,
          },
        ],
      };
  }
}

export const createProjectAction: Action = {
  name: 'CREATE_PROJECT',
  similes: [
    'GENERATE_CODE',
    'BUILD_PROJECT',
    'CREATE_PLUGIN',
    'CREATE_AGENT',
    'BUILD_PLUGIN',
    'BUILD_AGENT',
    'MAKE_PROJECT',
    'NEW_PROJECT',
    'DEVELOP_PLUGIN',
    'CODE_PROJECT',
  ],
  description:
    'Create a new ElizaOS project (plugin, agent, workflow, etc.) with AI-generated code',

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const text = message.content.text?.toLowerCase() || '';

    // Check for project creation keywords
    const createKeywords = ['create', 'build', 'make', 'generate', 'develop', 'code'];
    const projectKeywords = ['project', 'plugin', 'agent', 'bot', 'workflow', 'app', 'application'];

    const hasCreateKeyword = createKeywords.some((keyword) => text.includes(keyword));
    const hasProjectKeyword = projectKeywords.some((keyword) => text.includes(keyword));

    // Also check for "I want to..." patterns
    const wantPatterns = [
      /i want to (?:create|build|make)/i,
      /i need (?:a|an) (?:new )?(?:plugin|agent|bot)/i,
      /can you (?:create|build|make|help me)/i,
      /help me (?:create|build|make)/i,
    ];

    const hasWantPattern = wantPatterns.some((pattern) => pattern.test(text));

    return (hasCreateKeyword && hasProjectKeyword) || hasWantPattern;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: any,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult> => {
    elizaLogger.info('CREATE_PROJECT action triggered');

    const formsService = runtime.getService('forms') as FormsService | null;
    if (!formsService) {
      elizaLogger.error('Forms service not available');
      if (callback) {
        await callback({
          text: "I can't create projects right now - the forms service isn't available. Please try again later.",
          error: true,
        });
      }
      return {
        success: false,
        error: 'Forms service not available',
      };
    }

    const projectType = extractProjectType(message.content.text || '');
    const suggestedName = extractProjectName(message.content.text || '');

    elizaLogger.info(`Detected project type: ${projectType}, suggested name: ${suggestedName}`);

    // Create form for project details
    const formTemplate = getFormTemplate(projectType);

    const projectId = uuidv4();
    const formId = uuidv4() as UUID; // Separate UUID for the form
    const formData = {
      id: formId,
      ...formTemplate,
      onCancel: {
        action: 'CANCEL_FORM',
        data: { formId },
      },
      onSubmit: {
        action: 'GENERATE_CODE',
        data: {
          projectId,
          projectType,
        },
      },
    };

    // Create the form
    await formsService.createForm(formData);

    // Store project metadata
    const projectMetadata: ProjectMetadata = {
      id: projectId as UUID,
      name: suggestedName || 'Unnamed Project',
      type: projectType,
      description: message.content.text || '',
      createdAt: Date.now(),
      updatedAt: Date.now(),
      status: 'planning',
    };

    await runtime.createMemory(
      {
        id: uuidv4() as UUID,
        entityId: runtime.agentId,
        roomId: message.roomId,
        agentId: runtime.agentId,
        content: {
          text: `Project metadata: ${JSON.stringify(projectMetadata)}`,
          type: 'project_metadata',
          metadata: projectMetadata,
        },
        createdAt: Date.now(),
      },
      'project_metadata'
    );

    // Send response with form
    const responseText =
      projectType === 'agent'
        ? "Great! Let's create your ElizaOS agent. I'll help you design their personality and capabilities."
        : projectType === 'plugin'
          ? "Excellent! Let's build your ElizaOS plugin. I'll help you define its features and functionality."
          : "Perfect! Let's create your project. Please fill out the details below.";

    if (callback) {
      await callback({
        text: responseText,
        form: formData,
      });
    }

    return {
      success: true,
      data: {
        projectId,
        projectType,
        formId: formData.id,
      },
    };
  },

  examples: [
    [
      {
        name: '{{user1}}',
        content: {
          text: "I want to create a new ElizaOS plugin called 'weather-bot' that fetches weather data",
        },
      },
      {
        name: '{{agent}}',
        content: {
          text: "Excellent! Let's build your ElizaOS plugin. I'll help you define its features and functionality.",
          action: 'CREATE_PROJECT',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: {
          text: 'Can you help me build an agent for customer support?',
        },
      },
      {
        name: '{{agent}}',
        content: {
          text: "Great! Let's create your ElizaOS agent. I'll help you design their personality and capabilities.",
          action: 'CREATE_PROJECT',
        },
      },
    ],
  ],
};

File: cpp_merge/classified/packages/plugin-autocoder/src/actions/create-project.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/plugin-autocoder/src/actions/create-project.h"

any extractProjectName(string text)
{
    auto quotedMatch = text->match((new RegExp(std:("["']([^"']+)["'"))));
    if (quotedMatch) {
        return (*const_(quotedMatch))[1];
    }
    auto namedMatch = text->match((new RegExp(std:("(?:called|named)\s+(\S+)"))));
    if (namedMatch) {
        return (*const_(namedMatch))[1];
    }
    auto patterns = array<std::shared_ptr<RegExp>>{ (new RegExp(std:("create\s+(?:a\s+)?(?:new\s+)?(\S+)"))), (new RegExp(std:("(?:build|make)\s+(?:a\s+)?(\S+)"))) };
    for (auto& pattern : patterns)
    {
        auto match = text->match(pattern);
        if (match) {
            return (*const_(match))[1];
        }
    }
    return nullptr;
};


any getFormTemplate(string projectType)
{
    static switch_type __switch1650_6261 = {
        { any(std:("plugin")), 1 },
        { any(std:("agent")), 2 },
        { any(std:("workflow")), 3 }
    };
    switch (__switch1650_6261[projectType])
    {
    case 1:
        return object{
            object::pair{std:("title"), std:("ElizaOS Plugin Project")}, 
            object::pair{std:("description"), std:("Let's gather information to create your plugin")}, 
            object::pair{std:("fields"), array<object>{ object{
                object::pair{std:("id"), std:("project_name")}, 
                object::pair{std:("label"), std:("Project Name")}, 
                object::pair{std:("type"), std:("text")}, 
                object::pair{std:("placeholder"), std:("my-awesome-plugin")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("What would you like to name your plugin?")}
            }, object{
                object::pair{std:("id"), std:("description")}, 
                object::pair{std:("label"), std:("Description")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("A plugin that does amazing things...")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("Describe what your plugin will do")}
            }, object{
                object::pair{std:("id"), std:("requirements")}, 
                object::pair{std:("label"), std:("Requirements")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("List the key features and functionality...")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("What specific features or capabilities should this plugin have?")}
            }, object{
                object::pair{std:("id"), std:("apis")}, 
                object::pair{std:("label"), std:("External APIs (optional)")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("OpenAI API, Discord API, etc.")}, 
                object::pair{std:("required"), false}, 
                object::pair{std:("helpText"), std:("List any external APIs or services this plugin will use")}
            }, object{
                object::pair{std:("id"), std:("test_scenarios")}, 
                object::pair{std:("label"), std:("Test Scenarios (optional)")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("User asks for weather, plugin responds with...")}, 
                object::pair{std:("required"), false}, 
                object::pair{std:("helpText"), std:("Describe some test scenarios to validate the plugin works correctly")}
            } }}
        };
    case 2:
        return object{
            object::pair{std:("title"), std:("ElizaOS Agent Project")}, 
            object::pair{std:("description"), std:("Let's design your agent's personality and capabilities")}, 
            object::pair{std:("fields"), array<object>{ object{
                object::pair{std:("id"), std:("agent_name")}, 
                object::pair{std:("label"), std:("Agent Name")}, 
                object::pair{std:("type"), std:("text")}, 
                object::pair{std:("placeholder"), std:("My Assistant")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("What's your agent's name?")}
            }, object{
                object::pair{std:("id"), std:("personality")}, 
                object::pair{std:("label"), std:("Personality & Bio")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("A helpful, friendly assistant who...")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("Describe your agent's personality and background")}
            }, object{
                object::pair{std:("id"), std:("capabilities")}, 
                object::pair{std:("label"), std:("Capabilities")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("Can help with coding, answer questions...")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("What should your agent be able to do?")}
            }, object{
                object::pair{std:("id"), std:("plugins")}, 
                object::pair{std:("label"), std:("Plugins to Include")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("twitter, discord, knowledge...")}, 
                object::pair{std:("required"), false}, 
                object::pair{std:("helpText"), std:("Which ElizaOS plugins should this agent use?")}
            } }}
        };
    case 3:
        return object{
            object::pair{std:("title"), std:("Workflow Project")}, 
            object::pair{std:("description"), std:("Define your workflow automation")}, 
            object::pair{std:("fields"), array<object>{ object{
                object::pair{std:("id"), std:("workflow_name")}, 
                object::pair{std:("label"), std:("Workflow Name")}, 
                object::pair{std:("type"), std:("text")}, 
                object::pair{std:("placeholder"), std:("data-processing-workflow")}, 
                object::pair{std:("required"), true}
            }, object{
                object::pair{std:("id"), std:("description")}, 
                object::pair{std:("label"), std:("Description")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("This workflow processes data and...")}, 
                object::pair{std:("required"), true}
            }, object{
                object::pair{std:("id"), std:("steps")}, 
                object::pair{std:("label"), std:("Workflow Steps")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("1. Fetch data\
2. Process\
3. Store results")}, 
                object::pair{std:("required"), true}, 
                object::pair{std:("helpText"), std:("List the steps in your workflow")}
            }, object{
                object::pair{std:("id"), std:("triggers")}, 
                object::pair{std:("label"), std:("Triggers")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("placeholder"), std:("On schedule, webhook, manual...")}, 
                object::pair{std:("required"), false}
            } }}
        };
    default:
        return object{
            object::pair{std:("title"), std:("New Project")}, 
            object::pair{std:("description"), std:("Tell me about your project")}, 
            object::pair{std:("fields"), array<object>{ object{
                object::pair{std:("id"), std:("project_name")}, 
                object::pair{std:("label"), std:("Project Name")}, 
                object::pair{std:("type"), std:("text")}, 
                object::pair{std:("required"), true}
            }, object{
                object::pair{std:("id"), std:("description")}, 
                object::pair{std:("label"), std:("Description")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("required"), true}
            }, object{
                object::pair{std:("id"), std:("requirements")}, 
                object::pair{std:("label"), std:("Requirements")}, 
                object::pair{std:("type"), std:("textarea")}, 
                object::pair{std:("required"), true}
            } }}
        };
    }
};


std::shared_ptr<Action> createProjectAction = object{
    object::pair{std:("name"), std:("CREATE_PROJECT")}, 
    object::pair{std:("similes"), array<string>{ std:("GENERATE_CODE"), std:("BUILD_PROJECT"), std:("CREATE_PLUGIN"), std:("CREATE_AGENT"), std:("BUILD_PLUGIN"), std:("BUILD_AGENT"), std:("MAKE_PROJECT"), std:("NEW_PROJECT"), std:("DEVELOP_PLUGIN"), std:("CODE_PROJECT") }}, 
    object::pair{std:("description"), std:("Create a new ElizaOS project (plugin, agent, workflow, etc.) with AI-generated code")}, 
    object::pair{std:("validate"), [=](auto runtime, auto message, auto _state = undefined) mutable
    {
        shared text = OR((message->content->text->toLowerCase()), (string_empty));
        auto createKeywords = array<string>{ std:("create"), std:("build"), std:("make"), std:("generate"), std:("develop"), std:("code") };
        auto projectKeywords = array<string>{ std:("project"), std:("plugin"), std:("agent"), std:("bot"), std:("workflow"), std:("app"), std:("application") };
        auto hasCreateKeyword = createKeywords->some([=](auto keyword) mutable
        {
            return text["includes"](keyword);
        }
        );
        auto hasProjectKeyword = projectKeywords->some([=](auto keyword) mutable
        {
            return text["includes"](keyword);
        }
        );
        auto wantPatterns = array<std::shared_ptr<RegExp>>{ (new RegExp(std:("i want to (?:create|build|make)"))), (new RegExp(std:("i need (?:a|an) (?:new )?(?:plugin|agent|bot)"))), (new RegExp(std:("can you (?:create|build|make|help me)"))), (new RegExp(std:("help me (?:create|build|make)"))) };
        auto hasWantPattern = wantPatterns->some([=](auto pattern) mutable
        {
            return pattern->test(text);
        }
        );
        return OR(((AND((hasCreateKeyword), (hasProjectKeyword)))), (hasWantPattern));
    }
    }, 
    object::pair{std:("handler"), [=](auto runtime, auto message, auto state = undefined, auto _options = undefined, auto callback = undefined, auto _responses = undefined) mutable
    {
        elizaLogger->info(std:("CREATE_PROJECT action triggered"));
        auto formsService = as<any>(runtime->getService(std:("forms")));
        if (!formsService) {
            elizaLogger->error(std:("Forms service not available"));
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), std:("I can't create projects right now - the forms service isn't available. Please try again later.")}, 
                    object::pair{std:("error"), true}
                }); });
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), std:("Forms service not available")}
            };
        }
        auto projectType = extractProjectType(OR((message->content->text), (string_empty)));
        auto suggestedName = extractProjectName(OR((message->content->text), (string_empty)));
        elizaLogger->info(std:("Detected project type: ") + projectType + std:(", suggested name: ") + suggestedName + string_empty);
        auto formTemplate = getFormTemplate(projectType);
        auto projectId = uuidv4();
        auto formId = as<std::shared_ptr<UUID>>(uuidv4());
        auto formData = utils::assign(object{
            object::pair{std:("id"), formId}, 
            object::pair{std:("onCancel"), object{
                object::pair{std:("action"), std:("CANCEL_FORM")}, 
                object::pair{std:("data"), object{
                    object::pair{std:("formId"), std:("formId")}
                }}
            }}, 
            object::pair{std:("onSubmit"), object{
                object::pair{std:("action"), std:("GENERATE_CODE")}, 
                object::pair{std:("data"), object{
                    object::pair{std:("projectId"), std:("projectId")}, 
                    object::pair{std:("projectType"), std:("projectType")}
                }}
            }}
        }, formTemplate);
        std::async([=]() { formsService["createForm"](formData); });
        auto projectMetadata = object{
            object::pair{std:("id"), as<std::shared_ptr<UUID>>(projectId)}, 
            object::pair{std:("name"), OR((suggestedName), (std:("Unnamed Project")))}, 
            object::pair{std:("type"), projectType}, 
            object::pair{std:("description"), OR((message->content->text), (string_empty))}, 
            object::pair{std:("createdAt"), Date->now()}, 
            object::pair{std:("updatedAt"), Date->now()}, 
            object::pair{std:("status"), std:("planning")}
        };
        std::async([=]() { runtime->createMemory(object{
            object::pair{std:("id"), as<std::shared_ptr<UUID>>(uuidv4())}, 
            object::pair{std:("entityId"), runtime->agentId}, 
            object::pair{std:("roomId"), message->roomId}, 
            object::pair{std:("agentId"), runtime->agentId}, 
            object::pair{std:("content"), object{
                object::pair{std:("text"), std:("Project metadata: ") + JSON->stringify(projectMetadata) + string_empty}, 
                object::pair{std:("type"), std:("project_metadata")}, 
                object::pair{std:("metadata"), projectMetadata}
            }}, 
            object::pair{std:("createdAt"), Date->now()}
        }, std:("project_metadata")); });
        auto responseText = (projectType == std:("agent")) ? any(std:("Great! Let's create your ElizaOS agent. I'll help you design their personality and capabilities.")) ((projectType == std:("plugin")) ? std:("Excellent! Let's build your ElizaOS plugin. I'll help you define its features and functionality.") : std:("Perfect! Let's create your project. Please fill out the details below."));
        if (callback) {
            std::async([=]() { callback(object{
                object::pair{std:("text"), responseText}, 
                object::pair{std:("form"), formData}
            }); });
        }
        return object{
            object::pair{std:("success"), true}, 
            object::pair{std:("data"), object{
                object::pair{std:("projectId"), std:("projectId")}, 
                object::pair{std:("projectType"), std:("projectType")}, 
                object::pair{std:("formId"), formData->id}
            }}
        };
    }
    }, 
    object::pair{std:("examples"), array<array<any>>{ array<object>{ object{
        object::pair{std:("name"), std:("{{user1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I want to create a new ElizaOS plugin called 'weather-bot' that fetches weather data")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{agent}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Excellent! Let's build your ElizaOS plugin. I'll help you define its features and functionality.")}, 
            object::pair{std:("action"), std:("CREATE_PROJECT")}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("{{user1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Can you help me build an agent for customer support?")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{agent}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Great! Let's create your ElizaOS agent. I'll help you design their personality and capabilities.")}, 
            object::pair{std:("action"), std:("CREATE_PROJECT")}
        }}
    } } }}
};

void Main(void)
{
}

MAIN
